### PR TITLE
Dynamic mountpoint detection -- new version

### PR DESCRIPTION
**Pull request by vojtechtrefny**
New version of previous pull request -- https://github.com/rhinstaller/blivet/pull/47

Changes:
 - MountsCache is now a singleton
 - It is not neccessary to add devices to cache from devicetree
 - Opening files using "with"
 - One line return in MountsCache.getMountpoint
 - tests

### CODE REVIEW

File: blivet/formats/fs.py [hunk 1]
```diff
@@ -1604,14 +1615,14 @@ def _setOptions(self, options):
 
     @property
     def free(self):
-        if self._mountpoint:
-            # If self._mountpoint is defined, it means this tmpfs mount
+        if self.systeMountpoint:
```
Comment (mulkieran) [inline blivet/formats/fs.py]: Watch out for spelling mistakes!..."systeMountpoint". pylint will definitely catch these kinds of things.

File: blivet/formats/fs.py [hunk 2]
```diff
@@ -582,6 +582,13 @@ def loadModule(self):
         # also need to update the list of supported filesystems.
         update_kernel_filesystems()
 
+    @property
+    def systemMountpoint(self):
+        if not self.exists:
+            return None
+
+        return mountsCache.getMountpoint(self.device)
```
Comment (mulkieran) [inline blivet/formats/fs.py]: I would like it slightly better if this had a comment/docstring, if the final line was:

```
return mountsCache.getMountpoint(self.device, getattr(self, "subvolspec", None))
```
and the overriding method in BTRFS class went away.

**Comment by mulkieran** [comment]
Other than minor inline comments, that looks good to me. And a very nice improvement to our filesystem code, too.

**Comment by vojtechtrefny** [comment]
I made few changes based on comments by mulkieran and bcl but after force pushing them most of the comments disappeared...

**Comment by mulkieran** [comment]
Still looks good to me. FWIW, I ran a quick profile on the md5_file() method on my /proc/mounts and it's the read() method that chews up the bulk of the time, with updating the hash a distant second. And if we check whether there has been a change this way we just can't get away from reading those bits. Reading it in smaller chunks makes no difference...reading it in one big step doubles the time it takes.
Hoping that under normal use this change makes no perceptible difference (fingers crossed).

**Comment by vpodzime** [comment]
Other than the neat-pick above this still looks good to me.

**Comment by vojtechtrefny** [comment]
Docstring added. Pushed.

**Pull request closed without merging**